File: quote/Projects/PluginModelIO/CommonSaver.vb
﻿Imports System.Data.SqlClient
Imports System.Reflection
Imports System.Transactions
Imports System.ComponentModel

Imports Model.Common

Imports DB.QuoteDataBase
Imports DB.QuoteDataBaseTableAdapters

Public Class CommonSaver

    Public Shared ReadOnly COMPUTATION_PROPERTIES_ID = -1
    Public Shared ReadOnly OTHER_PROPERTIES_ID = -2
    Public Shared ReadOnly CUSTOM_PROPERTIES_ID = -3
    Public Shared ReadOnly NOTE_PROPERTIES_ID = -4

    Public Shared Sub SaveNoteProperties(ByVal id As Integer, _
                       ByVal obj As Object)
        SaveProperties(id, NOTE_PROPERTIES_ID, obj, True)
    End Sub

    Public Shared Sub SaveCustomProperties(ByVal id As Integer, _
                        ByVal obj As Object)
        SaveProperties(id, CUSTOM_PROPERTIES_ID, obj, True)
    End Sub

    Public Shared Sub SaveOtherProperties(ByVal id As Integer, _
                       ByVal obj As Object, _
                       ByVal SaveAll As Boolean)
        SaveProperties(id, OTHER_PROPERTIES_ID, obj, SaveAll)
    End Sub

    Public Shared Sub SaveComputationProperties(ByVal id As Integer, _
                     ByVal obj As Object, _
                     ByVal SaveAll As Boolean)
        SaveProperties(id, COMPUTATION_PROPERTIES_ID, obj, SaveAll)
    End Sub

    Private Shared Sub SaveProperties(ByVal id As Integer, _
                    ByVal childId As Integer, _
                    ByVal obj As Object, _
                    ByVal SaveAll As Boolean)
        Dim props As PropertyInfo() = obj.GetType.GetProperties
        Dim adaptor As New DB.QuoteDataBaseTableAdapters._QuotePropertiesTableAdapter
        Dim pNonDisplayable As PropertyInfo = Nothing
        For Each p As PropertyInfo In props
            'dd_added 2/8/2012
            If Not IsNothing(obj.GetType().GetProperty("Subject").GetValue(obj, Nothing)) Then
                pNonDisplayable = obj.Subject.GetType.GetProperty(p.Name)
            End If
            'dd_added end
            If SaveAll = False Then
                If Not p.CanWrite Then
                    Continue For
                End If
            End If

            Dim cat As String = "Misc"
            With cat
                Dim oa As CategoryAttribute() = p.GetCustomAttributes(GetType(CategoryAttribute), False)
                If oa.Length > 0 Then
                    cat = oa(0).Category
                End If
            End With

            Dim desc As String = ""
            With desc
                Dim oa As DescriptionAttribute() = p.GetCustomAttributes(GetType(DescriptionAttribute), False)
                If oa.Length > 0 Then
                    desc = oa(0).Description
                End If
            End With

            Dim browsable As Boolean = True
            With browsable
                Dim oa As BrowsableAttribute() = p.GetCustomAttributes(GetType(BrowsableAttribute), False)
                If oa.Length > 0 Then
                    browsable = oa(0).Browsable
                End If
            End With
            If Not browsable Then
                Continue For
            End If

            Dim s As String = Nothing
            Dim i As Integer = Nothing
            Dim d As Decimal = Nothing
            Dim b As Boolean = Nothing

            Dim o As Object
            'dd_changed 2/8/2012
            If IsNothing(pNonDisplayable) Then
                o = p.GetValue(obj, Nothing)
            Else
                o = pNonDisplayable.GetValue(obj.Subject, Nothing)
            End If
            'dd_changed end
            If TypeOf o Is Integer Then
                i = CInt(o)
                adaptor.Insert(id, childId, p.Name, Nothing, Nothing, i, cat, desc, Nothing)
            End If
            If TypeOf o Is String Then
                Dim t As New DB.QuoteDataBase._QuotePropertiesDataTable
                Dim max = t.PropertyStringValueColumn.MaxLength
                s = CStr(o)
                If s.Length > max Then
                    s = s.Substring(0, max - 1)
                End If
                adaptor.Insert(id, childId, p.Name, s, Nothing, Nothing, cat, desc, Nothing)
            End If
            If TypeOf o Is Decimal Then
                d = CDec(o)
                adaptor.Insert(id, childId, p.Name, Nothing, d, Nothing, cat, desc, Nothing)
            End If
            If TypeOf o Is DateTime Then
                Dim dt As DateTime = CDate(o)
                If dt.Year > 1 Then
                    adaptor.Insert(id, childId, p.Name, Nothing, Nothing, Nothing, cat, desc, dt)
                End If
            End If
            If TypeOf o Is Boolean Then
                b = CBool(o)
                If b Then
                    adaptor.Insert(id, childId, p.Name, "True", Nothing, Nothing, cat, desc, Nothing)
                Else
                    adaptor.Insert(id, childId, p.Name, "False", Nothing, Nothing, cat, desc, Nothing)
                End If
            End If
        Next
    End Sub

    Public Shared Sub DeleteComponents(ByVal id As Integer)
        Dim adaptor As New _QuoteDetailTableAdapter
        Dim table As DB.QuoteDataBase._QuoteDetailDataTable
        table = adaptor.GetDataByQuoteID(id)
        For Each row As _QuoteDetailRow In table.Rows
            adaptor.Delete(row.id)
        Next
    End Sub

    Private Shared Sub DeleteCustomProperties()
        DeleteProperties(CUSTOM_PROPERTIES_ID)
    End Sub

    Public Shared Sub DeleteProperties(ByVal QuoteID As Integer)

        Dim adaptor As New DB.QuoteDataBaseTableAdapters._QuotePropertiesTableAdapter
        Dim table As _QuotePropertiesDataTable = _
          adaptor.GetDataByQuoteID(QuoteID)
        For Each row As _QuotePropertiesRow In table.Rows
            adaptor.Delete(row.id)
        Next
    End Sub

    Public Shared Sub SaveComponents(ByVal q As Model.Template.Header, _
                     ByVal quoteId As Integer, _
                     ByVal SaveAll As Boolean)

        Dim adaptor As New _QuoteDetailTableAdapter
        Dim oldId As Integer = q.PrimaryProperties.CommonId
        Dim table As _QuoteDetailDataTable = adaptor.GetDataByQuoteID(oldId)
        For Each detail As Model.Common.Detail In q.Details
            adaptor.Connection.Open()
            Dim proxy As New DB.DetailTableProxy(adaptor)
            proxy.Transaction = adaptor.Connection.BeginTransaction
            'dd_Added sourceID, IsWire 10/3/11, SequenceNumber 10/7/11
            adaptor.Insert(quoteId, detail.SequenceNumber, _
                           detail.Qty, _
                           detail.Product.Code, _
                           detail.SourceID, _
                           detail.IsWire, detail.UnitOfMeasure)
            Dim cmd As SqlCommand = New SqlCommand( _
             "SELECT @@IDENTITY", adaptor.Connection)
            cmd.Transaction = proxy.Transaction
            Dim id As Integer = CInt(cmd.ExecuteScalar())
            proxy.Transaction.Commit()
            adaptor.Connection.Close()

            SaveProperties(quoteId, id, _
              detail.QuoteDetailProperties, SaveAll)
            detail.ClearDirty()
        Next

    End Sub


End Class

File: trunk/quote/Projects/PluginModelIO/CommonSaver.cs
namespace Model.IO
{
    using System;
    using System.ComponentModel;
    using System.Data.SqlClient;
    using System.Globalization;
    using System.Linq;
    using System.Reflection;
    
    using DB;
    using DB.QuoteDataBaseTableAdapters;
    
    using Model.Common;
    using Model.Template;

    public static class CommonSaver
    {
        public static readonly int ComputationPropertiesId = -1;
        public static readonly int OtherPropertiesId = -2;
        public static readonly int CustomPropertiesId = -3;
        public static readonly int NotePropertiesId = -4;
        
        public static void SaveNoteProperties(int id, SavableProperties obj)
        {
            SaveProperties(id, NotePropertiesId, obj, true);
        }

        public static void SaveCustomProperties(int id, SavableProperties obj)
        {
            SaveProperties(id, CustomPropertiesId, obj, true);
        }

        public static void SaveOtherProperties(int id, SavableProperties obj, bool saveAll)
        {
            SaveProperties(id, OtherPropertiesId, obj, saveAll);
        }

        public static void SaveComputationProperties(int id, SavableProperties obj, bool saveAll)
        {
            SaveProperties(id, ComputationPropertiesId, obj, saveAll);
        }

        public static void DeleteComponents(int id)
        {
            _QuoteDetailTableAdapter adaptor = new _QuoteDetailTableAdapter();
            DB.QuoteDataBase._QuoteDetailDataTable table = null;
            table = adaptor.GetDataByQuoteID(id);
            
            foreach (DB.QuoteDataBase._QuoteDetailRow row in table.Rows) 
            {
                adaptor.Delete(row.id);
            }
        }

        public static void DeleteProperties(int quoteId)
        {
            DB.QuoteDataBaseTableAdapters._QuotePropertiesTableAdapter adaptor = 
                new DB.QuoteDataBaseTableAdapters._QuotePropertiesTableAdapter();
            DB.QuoteDataBase._QuotePropertiesDataTable table = adaptor.GetDataByQuoteID(quoteId);
            
            foreach (DB.QuoteDataBase._QuotePropertiesRow row in table.Rows) 
            {
                adaptor.Delete(row.id);
            }
        }

        public static void SaveComponents(
            Model.Common.Header header, 
            int quoteId, 
            bool saveAll)
        {
            _QuoteDetailTableAdapter adaptor = new _QuoteDetailTableAdapter();
            
            foreach (Model.Common.Detail detail in header.Details) 
            {
                adaptor.Connection.Open();
                DB.DetailTableProxy proxy = new DB.DetailTableProxy(adaptor);
                proxy.Transaction = adaptor.Connection.BeginTransaction();
                
                adaptor.Insert(
                    quoteId, 
                    detail.SequenceNumber, 
                    detail.Qty, 
                    detail.Product.Code, 
                    detail.SourceId, 
                    detail.IsWire, 
                    detail.UnitOfMeasure);
                
                SqlCommand cmd = new SqlCommand("SELECT @@IDENTITY", adaptor.Connection);
                cmd.Transaction = proxy.Transaction;
                int id = Convert.ToInt32(cmd.ExecuteScalar(), CultureInfo.CurrentCulture);
                proxy.Transaction.Commit();
                adaptor.Connection.Close();

                SaveProperties(quoteId, id, detail.QuoteDetailProperties, saveAll);
                detail.IsDirty = false;
            }
        }
        
        private static void SaveProperties(
            int id, 
            int childId, 
            SavableProperties obj, 
            bool saveAll)
        {
            PropertyInfo[] props = obj.GetType().GetProperties();
            DB.QuoteDataBaseTableAdapters._QuotePropertiesTableAdapter adaptor = 
                new DB.QuoteDataBaseTableAdapters._QuotePropertiesTableAdapter();
            PropertyInfo nonDisplayable = null;
            
            IHasSubject hasSubject = obj as IHasSubject;
            foreach (PropertyInfo p in props) 
            {
                if (hasSubject != null)
                {
                    nonDisplayable = hasSubject.Subject.GetType().GetProperty(p.Name);
                }
                
                if (saveAll == false) 
                {
                    if (!p.CanWrite) 
                    {
                        continue;
                    }
                }

                string cat = "Misc";
                CategoryAttribute[] oa = 
                    p.GetCustomAttributes(typeof(CategoryAttribute), false) as CategoryAttribute[];
                if (oa.Length > 0) 
                {
                    cat = oa[0].Category;
                }

                string desc = string.Empty;
                DescriptionAttribute[] oa2 = 
                    p.GetCustomAttributes(typeof(DescriptionAttribute), false) as DescriptionAttribute[];
                if (oa2.Length > 0) 
                {
                    desc = oa2[0].Description;
                }

                bool browsable = true;
                BrowsableAttribute[] oa3 = 
                    p.GetCustomAttributes(typeof(BrowsableAttribute), false) as BrowsableAttribute[];
                if (oa3.Length > 0) 
                {
                    browsable = oa3[0].Browsable;
                }
                
                if (!browsable) 
                {
                    continue;
                }

                string s = null;
                int i;
                decimal d;
                bool b;

                object o = null;
                if (nonDisplayable == null)
                {
                    o = p.GetValue(obj, null);
                } 
                else
                {
                    o = nonDisplayable.GetValue(hasSubject.Subject, null);
                }
                
                if (o is int) 
                {
                    i = Convert.ToInt32(o, CultureInfo.CurrentCulture);
                    adaptor.Insert(id, childId, p.Name, null, null, i, cat, desc, null);
                }
                
                if (o is string) 
                {
                    DB.QuoteDataBase._QuotePropertiesDataTable t = 
                        new DB.QuoteDataBase._QuotePropertiesDataTable();
                    var max = t.PropertyStringValueColumn.MaxLength;
                    s = Convert.ToString(o, CultureInfo.CurrentCulture);
                    
                    if (s.Length > max) 
                    {
                        s = s.Substring(0, max - 1);
                    }
                    
                    adaptor.Insert(id, childId, p.Name, s, null, null, cat, desc, null);
                }
                
                if (o is decimal) 
                {
                    d = Convert.ToDecimal(o, CultureInfo.CurrentCulture);
                    adaptor.Insert(id, childId, p.Name, null, d, null, cat, desc, null);
                }
                
                if (o is DateTime) 
                {
                    DateTime dt = Convert.ToDateTime(o, CultureInfo.CurrentCulture);
                    if (dt.Year > 1) 
                    {
                        adaptor.Insert(id, childId, p.Name, null, null, null, cat, desc, dt);
                    }
                }
                
                if (o is bool) 
                {
                    b = Convert.ToBoolean(o, CultureInfo.CurrentCulture);
                    if (b) 
                    {
                        adaptor.Insert(id, childId, p.Name, "True", null, null, cat, desc, null);
                    } 
                    else 
                    {
                        adaptor.Insert(id, childId, p.Name, "False", null, null, cat, desc, null);
                    }
                }
            }
        }
    }
}
